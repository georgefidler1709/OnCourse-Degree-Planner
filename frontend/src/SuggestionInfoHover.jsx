// Modified from:
// https://gist.github.com/lou/571b7c0e7797860d6c555a9fdc0496f9

/*
* Usage:
* <SuggestionInfoHover
*    content={<div>Holy guacamole! I'm Sticky.</div>}
*    placement="top"
*    onMouseEnter={() => { }}
*    delay={200}
* >
*   <div>Show the sticky tooltip</div>
* </SuggestionInfoHover>
*/

import React, {Component} from 'react'
import { Overlay, Popover } from 'react-bootstrap'

class SuggestionInfoHover extends Component {
  constructor(props) {
    super(props)

    this.handleMouseEnter = this.handleMouseEnter.bind(this)
    this.handleMouseLeave = this.handleMouseLeave.bind(this)

    this.state = {
      showPopover: false,
    }
  }

  handleMouseEnter() {
    const {delay} = this.props

    this.setTimeoutConst = setTimeout(() => {
      this.setState({ showPopover: true }, () => {
      })
    }, delay);
  }

  handleMouseLeave() {
    clearTimeout(this.setTimeoutConst)
    this.setState({ showPopover: false })
  }

  componentWillUnmount() {
    if (this.setTimeoutConst) {
      clearTimeout(this.setTimeoutConst)
    }
  }

  render() {
    let { content, children, placement } = this.props

    const child = React.Children.map(children, (child) => (
      React.cloneElement(child, {
        onMouseEnter: this.handleMouseEnter,
        onMouseLeave: this.handleMouseLeave,
        ref: (node) => {
          this._child = node
          const { ref } = child
          if (typeof ref === 'function') {
            ref(node);
          }
        }
      })
    ))[0]

    return(
      <React.Fragment>
        {child}
        <Overlay
          show={this.state.showPopover}
          placement={placement}
          target={this._child}
          shouldUpdatePosition={true}
        >
          <Popover
            onMouseEnter={() => {
              this.setState({ showPopover: true })
            }}
            onMouseLeave={this.handleMouseLeave}
            id='popover'
          >
            <Popover.Content>{content}</Popover.Content>
          </Popover>
        </Overlay>
      </React.Fragment>
    )
  }
}

export default SuggestionInfoHover