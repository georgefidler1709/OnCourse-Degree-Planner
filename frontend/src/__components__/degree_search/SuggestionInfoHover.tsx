// Modified from:
// https://gist.github.com/lou/571b7c0e7797860d6c555a9fdc0496f9

/*
* Usage:
* <SuggestionInfoHover
*    content={<div>Holy guacamole! I'm Sticky.</div>}
*    placement="top"
*    delay={200}
* >
*   <div>Show the sticky tooltip</div>
* </SuggestionInfoHover>
*/

import React, {Component, RefObject} from 'react'
import { Overlay, Popover } from 'react-bootstrap'
import { Position } from '../../Types'


interface SuggestionInfoHoverProps { 
  content: JSX.Element
  placement?: Position
  delay: number
  children: JSX.Element
  key: string
}

class SuggestionInfoHover extends Component<SuggestionInfoHoverProps, {showPopover: boolean}> {
  private setTimeoutConst?: ReturnType<typeof setTimeout>
  private targetRef: RefObject<Component>

  constructor(props: SuggestionInfoHoverProps) {
    super(props)

    this.handleMouseEnter = this.handleMouseEnter.bind(this)
    this.handleMouseLeave = this.handleMouseLeave.bind(this)
    this.targetRef = React.createRef()

    this.state = {
      showPopover: false,
    }
  }

  handleMouseEnter() {
    const {delay} = this.props

    this.setTimeoutConst = setTimeout(() => {
      this.setState({ showPopover: true }, () => {
      })
    }, delay);
  }

  handleMouseLeave() {
    if(this.setTimeoutConst) {
      clearTimeout(this.setTimeoutConst)
    }
    this.setState({ showPopover: false })
  }

  componentWillUnmount() {
    if (this.setTimeoutConst) {
      clearTimeout(this.setTimeoutConst)
    }
  }

  render() {
    let { content, children, placement } = this.props

    return(
      <React.Fragment>
        {React.cloneElement(children, {
          onMouseEnter: this.handleMouseEnter,
          onMouseLeave: this.handleMouseLeave,
          ref: this.targetRef
        })}
        <Overlay
          show={this.state.showPopover}
          placement={placement}
          target={this.targetRef.current!}
        >
          <Popover
            onMouseEnter={() => {
              this.setState({ showPopover: true })
            }}
            onMouseLeave={this.handleMouseLeave}
            id='popover'
          >
            <Popover.Content>{content}</Popover.Content>
          </Popover>
        </Overlay>
      </React.Fragment>
    )
  }
}

export default SuggestionInfoHover
