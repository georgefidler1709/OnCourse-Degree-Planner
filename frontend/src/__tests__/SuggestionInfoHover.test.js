import React from 'react'
import { shallow} from 'enzyme';
import SuggestionInfoHover from '../__components__/degree_search/SuggestionInfoHover';
import { Popover } from 'react-bootstrap'


describe('Rendering a degree tooltip', () => {
    it('renders correctly', () => {
      let wrapper = shallow(
        <SuggestionInfoHover
          content={<div>More Info</div>}
          placement="top"
          delay={200}
        >
          <div className="test">Show tooltip</div>
        </SuggestionInfoHover>
      )
      expect(wrapper).toMatchSnapshot();
    });

    it('displays more info on mouseover', () => {
      let wrapper = shallow(
        <SuggestionInfoHover
          content={<div>More Info</div>}
          placement="top"
          delay={200}
        >
          <div className="test">Show tooltip</div>
        </SuggestionInfoHover>
      )

      wrapper.find('.test').simulate('mouseover');
      wrapper.update();

      expect(wrapper.find(Popover.Content).text()).toBe('More Info')
      expect(wrapper).toMatchSnapshot();
    });


});
