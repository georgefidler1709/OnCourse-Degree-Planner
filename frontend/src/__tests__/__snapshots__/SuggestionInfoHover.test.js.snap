// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering a degree tooltip displays more info on mouseover 1`] = `
<Fragment>
  <div
    className="test"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Show tooltip
  </div>
  <Overlay
    placement="top"
    rootClose={false}
    show={false}
    target={null}
    transition={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "appear": false,
          "in": false,
          "mountOnEnter": false,
          "timeout": 300,
          "unmountOnExit": false,
        },
        "displayName": "Fade",
        "render": [Function],
      }
    }
  >
    <Bootstrap(Popover)
      id="popover"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <ForwardRef>
        <div>
          More Info
        </div>
      </ForwardRef>
    </Bootstrap(Popover)>
  </Overlay>
</Fragment>
`;

exports[`Rendering a degree tooltip renders correctly 1`] = `
<Fragment>
  <div
    className="test"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Show tooltip
  </div>
  <Overlay
    placement="top"
    rootClose={false}
    show={false}
    target={null}
    transition={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "appear": false,
          "in": false,
          "mountOnEnter": false,
          "timeout": 300,
          "unmountOnExit": false,
        },
        "displayName": "Fade",
        "render": [Function],
      }
    }
  >
    <Bootstrap(Popover)
      id="popover"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <ForwardRef>
        <div>
          More Info
        </div>
      </ForwardRef>
    </Bootstrap(Popover)>
  </Overlay>
</Fragment>
`;
